raft/storage.go

```diff
@@ -57,6 +57,17 @@ type Storage interface {
 	Snapshot() (pb.Snapshot, error)
 }
 
+// InitialState is the return value of the Storage.InitialState method.
+type InitialState struct {
+	HardState pb.HardState
+	ConfState pb.ConfState
+
+	// AppliedIndex is the last index which has been applied to the state machine,
+	// for implementations which persist the state machine to disk. If the state
+	// machine is not persistent, this field may be zero.
+	AppliedIndex uint64
```

Comment:
For us, the applier and storage are the same, and while we might separate the implementations in the future, I think we'd always prefer to make the Storage object know about the applier rather than pass the applied index in separately. We've been moving away from the arguments to newRaft and making the Storage object entirely self-contained (e.g. https://github.com/cockroachdb/cockroach/commit/48237bd242fee60869d94653b82c5d046f0e7c7b)